packages/telemetry/src/index.ts

```diff
@@ -112,39 +143,76 @@ export class Telemetry implements TelemetryApi {
     name,
     attachType = 'child',
     active = false,
+    parentSpan,
     opts = {},
+    key,
+    isVerbose = false,
   }: startSpanOptions) {
     // Currently the latest span replaces any previous open or closed span and you can no longer access the replaced span.
     // This works well enough for now but may cause issue in the future.
 
+    // if the span is declared in verbose mode, but verbosity is disabled, no-op the span creation
+    if (isVerbose && !this.isVerbose) {
+      return undefined
+    }
+
     let span: Span
 
-    // If root or implied root
-    if (attachType === 'root' || this.activeSpanQueue.length < 1) {
+    let parent: Span | undefined
+
+    if (attachType === 'root' || (this.activeSpanQueue.length < 1 && !parentSpan)) {
       if (this.rootContext) {
         // Start span with external context
         span = this.tracer.startSpan(name, opts, this.rootContext)
+
+        // This can only apply attributes set on the external root set up until the point at which it was sent.
+        if (this.rootAttributes) {
+          span.setAttributes(this.rootAttributes)
+        }
       } else {
         // Start span with no context
         span = this.tracer.startSpan(name, opts)
       }
     } else { // attach type must be child
+      // Prefer passed in parent
+      parent = parentSpan || this.activeSpanQueue[this.activeSpanQueue.length - 1]
+
       // Create a context from the active span.
-      const ctx = openTelemetry.trace.setSpan(openTelemetry.context.active(), this.activeSpanQueue[this.activeSpanQueue.length - 1]!)
+      const ctx = openTelemetry.trace.setSpan(openTelemetry.context.active(), parent!)
 
       // Start span with parent context.
       span = this.tracer.startSpan(name, opts, ctx)
     }
 
-    // Save off span, duplicate names currently not handled.
-    this.spans[name] = span
+    // Save off span, keys must be unique.
+    const spanKey = key || name
 
-    // If this is an active span, set it as the new active span
-    if (active) {
-      const _end = span.end
+    this.spans[spanKey] = span
+
+    // Setup function on span to recursively get parent attributes.
+    // Not bothering with types here since we only need this function within this function.
+    // @ts-ignore
+    span.getAllAttributes = () => {
+      // @ts-ignore
+      const parentAttributes = parent && parent.getAllAttributes ? parent.getAllAttributes() : {}
+
+      const allAttributes = {
+        // @ts-ignore
+        ...span.attributes,
+        ...parentAttributes,
+      }
+
+      // never propagate name
+      delete allAttributes['name']
 
-      // override the end function to allow us to pop the span off the queue if found.
-      span.end = (endTime) => {
+      return (allAttributes)
```

Comment:
any idea why we need this in parenthesis?

Reply:
no reason